sage/ip_lookup.py

```diff
@@ -0,0 +1,105 @@
+"""
+~40,000 distinct ip addresses recorded "downloading a file" in Q1 of 2024
+Using http://ip-api.com/batch API, I was able to extract the information of all ~40000 IP addresses.
+This information contains city, country, region, etc which is then stored back into snowflake
+for analysis
+
+This eventually could be an Airflow DAG, so not going spending too much time on it
+"""
+import requests
+import time
+
+from dotenv import dotenv_values
+import snowflake.connector
+from snowflake.connector.pandas_tools import write_pandas
+import pandas as pd
+import numpy as np
+
+
+def get_ip_info(ip_list: list) -> dict:
+    """Get IP information from ip-api.com: http://ip-api.com/batch
+    
+    Args:
+        ip_list (list): list of IP addresses
+    
+    Returns:
+        dict: IP information like city, country, region, lat, long, asn, etc
+    """
+    # HACK this is not great...
+    while True:
+        try:
+            ip_info_response = requests.post('http://ip-api.com/batch', json=ip_list)
+        except Exception as err:
+            print(ip_info_response.status_code)
+            if ip_info_response.status_code == 429:
```

Comment:
Future change would be to use the backoff package.

https://pypi.org/project/backoff/